File: visitor/cpp/visitor.cc
#include "visitor/cpp/visitor.h"
#include <iostream>

void CustomerCol::add(Customer *c) {
    customers.push_back(c);
}

void CustomerCol::accept(Visitor *v) {
    for(auto c : customers) {
        v->visit(c);
    }
}

void Customer::accept(Visitor *v) {
    v->visit(this);
}

EnterpriseCustomer::EnterpriseCustomer(const std::string &name_):name(name_) {}

// void EnterpriseCustomer::accept(Visitor *v) {
//     v->visit(this);
// }

std::string EnterpriseCustomer::get_name() {
    return this->name;
}

EnterpriseCustomer::~EnterpriseCustomer() = default;

IndividualCustomer::IndividualCustomer(const std::string &name_):name(name_) {}

// void IndividualCustomer::accept(Visitor *v) {
//     v->visit(this);
// }

std::string IndividualCustomer::get_name() {
    return this->name;
}

IndividualCustomer::~IndividualCustomer() = default;

void ServiceRequestVisitor::visit(Customer *c) {
    if(auto ec = dynamic_cast<EnterpriseCustomer*>(c)) {
        std::cout << "serving enterprise customer " << ec->get_name() << "\n";
        return;
    }
    if(auto ic = dynamic_cast<IndividualCustomer*>(c)) {
        std::cout << "serving individual customer " << ic->get_name() << "\n";
        return;
    }
}

void AnalysisVisitor::visit(Customer *c) {
    if(auto ec = dynamic_cast<EnterpriseCustomer*>(c)) {
        // fmt.Printf("serving enterprise customer %s\n", c.name)
        std::cout << "analysis enterprise customer " << ec->get_name() << "\n";
        return;
    }
}

File: visitor/visitor.go
package visitor

import (
	"fmt"
)

type Visitor interface {
	Visit(Custom)
}

type Custom interface {
	Accept(Visitor)
}

type CustomerCol struct {
	customers []Custom
}

func (col *CustomerCol) Add(c Custom) {
	col.customers = append(col.customers, c)
}

func (col *CustomerCol) Accept(v Visitor) {
	for _, custom := range col.customers {
		custom.Accept(v)
	}
}

type baseCustomer struct {
	name string
}

func (bc baseCustomer) Accept(v Visitor) {
	v.Visit(bc)
}

type EnterpriseCustomer struct {
	*baseCustomer
}

func NewEnterpriseCustomer(name string) *EnterpriseCustomer {
	return &EnterpriseCustomer{
		baseCustomer: &baseCustomer{
			name: name,
		},
	}
}

func (ec *EnterpriseCustomer) Accept(v Visitor) {
	v.Visit(ec)
}

type IndividualCustomer struct {
	*baseCustomer
}

func NewIndividualCustomer(name string) *IndividualCustomer {
	return &IndividualCustomer{
		baseCustomer: &baseCustomer{
			name: name,
		},
	}
}

func (ic *IndividualCustomer) Accept(v Visitor) {
	v.Visit(ic)
}

type AnalysisVisitor struct{}

type ServiceRequestVisitor struct{}

func (*AnalysisVisitor) Visit(custom Custom) {
	switch c := custom.(type) {
	case *EnterpriseCustomer:
		fmt.Printf("analysis enterprise customer %s\n", c.name)
	}
}

func (*ServiceRequestVisitor) Visit(custom Custom) {
	switch c := custom.(type) {
	case *EnterpriseCustomer:
		fmt.Printf("serving enterprise customer %s\n", c.name)
	case *IndividualCustomer:
		fmt.Printf("serving individual customer %s\n", c.name)
	}
}
